sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,188 @@
 use std::env;
 use std::fs;
 use std::path::Path;
+use std::path::PathBuf;
 
-use rom_ext_config::parser::ParsedConfig;
-use rom_ext_image::image::RawImage;
+use rom_ext_image::image::Image;
+use rom_ext_image::manifest;
 
 use mundane::hash::Sha256;
 use mundane::public::rsa::RsaPkcs1v15;
-use mundane::public::rsa::RsaPrivKey;
-use mundane::public::rsa::RsaSignature;
 use mundane::public::rsa::B3072;
 use mundane::public::DerPrivateKey;
 use mundane::public::Signature;
 
-fn main() {
-    let arg: String = env::args().nth(1).expect("Config path is missing");
+use anyhow::bail;
+use anyhow::ensure;
+use anyhow::Context;
+use anyhow::Result;
 
-    let config_path = Path::new(&arg);
+// Type aliases for convenience.
+type ImageSignature =
+    mundane::public::rsa::RsaSignature<B3072, RsaPkcs1v15, Sha256>;
+type PrivateKey = mundane::public::rsa::RsaPrivKey<B3072>;
 
-    // Parse the config.
-    let config = ParsedConfig::new(&config_path);
-
-    // Read raw binary.
-    let mut image = RawImage::new(&config.input_files.image_path);
-
-    // Get the private key used for signature generation. It is also used to
-    // extract key public exponent and modulus.
-    let private_key_der = fs::read(&config.input_files.private_key_der_path)
-        .expect("Failed to read the image!");
-
-    // Update "signed" manifest fields.
-    image.update_static_fields(&config);
-
-    // Convert ASN.1 DER private key into Mundane RsaPrivKey.
-    let private_key = RsaPrivKey::parse_from_der(&private_key_der)
-        .expect("Failed to parse private key!");
-
-    let mut exponent = private_key.public_exponent_be();
-    // Encode public key components in little-endian byte order.
-    exponent.reverse();
-    image.update_exponent_field(&exponent);
-
-    let mut modulus = private_key.public_modulus_be();
-    // Encode public key components in little-endian byte order.
-    modulus.reverse();
-    image.update_modulus_field(&modulus);
-
-    // Produce the signature from concatenated system_state_value,
-    // device_usage_value and the portion of the "signed" portion of the image.
-    image.update_timestamp_field();
-    let image_sign_data = image.data_to_sign();
-    let device_usage_value =
-        &device_usage_value(&config.input_files.usage_constraints_path);
-    let system_state_value =
-        &system_state_value(&config.input_files.system_state_value_path);
+struct Args {
+    input: PathBuf,
+    priv_key: PathBuf,
+    output: PathBuf,
+}
 
-    let mut message_to_sign = Vec::<u8>::new();
-    message_to_sign.extend_from_slice(system_state_value);
-    message_to_sign.extend_from_slice(device_usage_value);
-    message_to_sign.extend_from_slice(image_sign_data);
+impl Args {
+    pub fn new(args: env::Args) -> Result<Args> {
```

Comment:
I think you probably want ArgsOs here.

Reply:
Done (not sure if strictly necessary though, since I would expect these arguments to be valid anyway).